### PR TITLE
Add OnGetFlows and OnFlowDelivery

### PR DESCRIPTION
**Pull request by tgraf**
Depends on #213

### CODE REVIEW

File: pkg/server/local_observer.go [hunk 1]
```diff
@@ -220,12 +220,20 @@ func (s *LocalObserverServer) GetFlows(
 	req *observer.GetFlowsRequest,
 	server observer.Observer_GetFlowsServer,
 ) (err error) {
+	ctx := server.Context()
```
Comment (gandro) [inline pkg/server/local_observer.go:223]: I think we should actually wrap this context here in a `WithCancel` and call cancel when `GetFlows` returns, to ensure that the context lives only as as long as the request itself. This allows implementors of the `OnGetFlows` hook to release any resources it allocated for this particular request.

We probably also want to pass this per-request-context down to the `OnBuildFilter` and `OnFlowDelivery` hooks as well.

File: pkg/server/local_observer.go [hunk 2]
```diff
@@ -259,7 +267,7 @@ func (s *LocalObserverServer) GetFlows(
 
 nextFlow:
 	for ; ; i++ {
-		flow, err := flowsReader.Next(server.Context())
+		flow, err := flowsReader.Next(ctx)
```
Comment (gandro) [inline pkg/server/local_observer.go:270]: While not introduced by this PR, I just notice it now: Passing down `server.Context()` to the flow reader is wrong. I believe this is another reason we want introduce a per-request context (as stated above).

To demonstrate: The flow reader spawns goroutines to implement `follow` requests. These goroutines must be terminated once the request is finished, which is achieved via context cancellation. If we pass down `server.Context()` here, these per-request goroutines will be kept alive until the server is terminated, which is essentially a goroutine leak.  The context passed down to `flowsReader.Next` must only live as long as the request.

File: pkg/server/local_observer_test.go [hunk 3]
```diff
@@ -231,3 +231,64 @@ func TestLocalObserverServer_OnFlowDelivery(t *testing.T) {
...
+		FakeGRPCServerStream: &FakeGRPCServerStream{
+			OnContext: func() context.Context {
+				return context.Background()
+			},
+		},
+	}
+
+	var key int
+	onGetFlows := func(ctx context.Context, req *observer.GetFlowsRequest) (context.Context, error) {
+		return context.WithValue(ctx, key, 10), nil
```
Comment (gandro) [inline pkg/server/local_observer_test.go:254]: See `golint`: `should not use basic type int as key in context.WithValue`

Introducing a `type ctxKey int` should be sufficient to satisfy the lint.

File: pkg/server/local_observer.go (hunk 1 above)
Comment (gandro) [reply pkg/server/local_observer.go:223]: Update: I have fixed the context in #216.

File: pkg/server/local_observer_test.go (hunk 3 above)
Comment (tgraf) [reply pkg/server/local_observer_test.go:254]: Yep, this will need a type we can use for this consistently.

**Comment by tgraf** [comment]
Closing in favor of https://github.com/cilium/cilium/pull/10896

**Pull request closed without merging**